autogen/function_utils.py

```diff
@@ -0,0 +1,197 @@
+    """Get the required parameters of a function
+
+    Args:
+        signature: The signature of the function as returned by inspect.signature
+
+    Returns:
+        A list of the required parameters of the function
+    """
+    return [k for k, v in typed_signature.parameters.items() if v.default == inspect.Signature.empty]
+
+
+def get_parameters(required: List[str], param_annotations: Dict[str, Union[Annotated[Type, str], Type]]) -> Parameters:
+    """Get the parameters of a function as defined by the OpenAI API
+
+    Args:
+        required: The required parameters of the function
+        hints: The type hints of the function as returned by typing.get_type_hints
+
+    Returns:
+        A Pydantic model for the parameters of the function
+    """
+    return Parameters(
+        properties={k: get_parameter_json_schema(k, v) for k, v in param_annotations.items() if k != "return"},
+        required=required,
+    )
+
+
+def get_function_schema(f: Callable[..., Any], *, name: Optional[str] = None, description: str) -> Dict[str, Any]:
+    """Get a JSON schema for a function as defined by the OpenAI API
+
+    Args:
+        f: The function to get the JSON schema for
+        name: The name of the function
+        description: The description of the function
+
+    Returns:
+        A JSON schema for the function
+
+    Raises:
+        TypeError: If the function is not annotated
+
+    Examples:
+        ```
+        def f(a: Annotated[str, "Parameter a"], b: int = 2, c: Annotated[float, "Parameter c"] = 0.1) -> None:
+            pass
+
+        get_function_schema(f, description="function f")
+
+        #   {'type': 'function',
+        #    'function': {'description': 'function f',
+        #        'name': 'f',
+        #        'parameters': {'type': 'object',
+        #           'properties': {'a': {'type': 'str', 'description': 'Parameter a'},
+        #               'b': {'type': 'int', 'description': 'b'},
+        #               'c': {'type': 'float', 'description': 'Parameter c'}},
+        #           'required': ['a']}}}
+            ```
+
+    """
+    typed_signature = get_typed_signature(f)
+    param_annotations = {k: v.annotation for k, v in typed_signature.parameters.items()}
+    return_annotation = get_typed_return_annotation(f)
+    missing_annotations = [k for k, v in param_annotations.items() if v is inspect.Signature.empty]
+
+    if return_annotation is None:
+        raise TypeError(
+            "The return type of a function must be annotated as either 'str', a subclass of "
+            + "'pydantic.BaseModel' or an union of the previous ones."
+        )
+
+    if missing_annotations != []:
+        [f"'{k}'" for k in missing_annotations]
+        raise TypeError(
+            f"All parameters of a function '{f.__name__}' must be annotated. "
+            + "The annotations are missing for parameters: {', '.join(missing)}"
+        )
```

Comment:
One disadvantage of the decorator method is that it doesn't work with a function w/o complete annotations while the current json way does.

What about we only raise errors for parameters w/o default values? For others, we log a warning message instead of raise an error?

Reply:
Yes, that's much better. Done in the new commit.